ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,351 @@
+    SendHeartbeatTask sendTask = new SendHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(sendTask, 0, configs.heartbeatSendIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    CheckHeartbeatTask processTask = new CheckHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(processTask, 0, configs.heartbeatCheckIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    DiscoverClusterTask discoverTask = new DiscoverClusterTask();
+    scheduledExecutorService.scheduleAtFixedRate(discoverTask, 0, configs.discoverClusterIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+
+  }
+
+  void shutDown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  public ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> getReceivedHeartbeats() {
+    return receivedHeartbeats;
+  }
+
+  public void registerHeartbeat(HostInfo hostInfo, long timestamp) {
+    receivedHeartbeats.putIfAbsent(hostInfo, new TreeMap<>());
+    TreeMap<Long, HeartbeatInfo> heartbeats = receivedHeartbeats.get(hostInfo);
+
+    synchronized (heartbeats) {
+      heartbeats.put(timestamp, new HeartbeatInfo(timestamp));
+    }
+  }
+
+  ConcurrentHashMap<URI, Boolean> getRemoteHostsURI() {
+    return remoteHostsURI;
+  }
+
+  public Map<String, Boolean> getAliveHosts() {
+    return aliveHosts;
+  }
+
+  class CheckHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      long now = System.currentTimeMillis();
+      long windowStart = now - configs.heartbeatWindowMs;
+      processHeartbeats(windowStart, now);
+    }
+
+    @VisibleForTesting
+    void runWithWindow(long windowStart, long windowEnd) {
+      processHeartbeats(windowStart, windowEnd);
+    }
+
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      if (receivedHeartbeats.isEmpty()) {
+
+        remoteHostsInfo.forEach((host, bool) ->
+                                    aliveHosts.compute(host.toString(), (key, val) -> false));
+      }
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        synchronized (heartbeats) {
+          // 1. remove heartbeats older than window
+          while (!heartbeats.isEmpty()) {
+            Entry<Long, HeartbeatInfo> nestedEntry = heartbeats.firstEntry();
+            long timestamp = nestedEntry.getKey();
+            if (timestamp < windowStart) {
+              heartbeats.pollFirstEntry();
+            } else {
+              break;
+            }
+          }
+          // 2. count consecutive missed heartbeats and mark as up
+          downPolicy(host, windowStart, windowEnd, heartbeats);
+        }
+      }
+    }
+
+
+    private void downPolicy(final HostInfo host, final long windowStart, final long windowEnd,
```

Comment:
rename to `decideStatus` or `getHostStatus` returning boolean that you set in the map?